File: 2855-maximum-number-of-jumps-to-reach-the-last-index/maximum-number-of-jumps-to-reach-the-last-index.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var maximumJumps = function(nums, target) {
    const n = nums.length;
    const memo = new Array(n);
    let maxi = rec(0)

    return maxi === -Infinity ? -1 : maxi

    function rec(i) {

        if(i == n - 1) return 0;
        if(i >= n) return -Infinity

        if(memo[i] != undefined) return memo[i]

        // console.log(i)
        
        let max = -Infinity;
        for(let j = i + 1; j < n; j++) {

            const val = nums[j] - nums[i];
            if(-target <= val && val <= target) {
                max = Math.max(max, 1 + rec(j))
            }
        }

        return memo[i] = max;


    }
};

File: 2855-maximum-number-of-jumps-to-reach-the-last-index/maximum-number-of-jumps-to-reach-the-last-index.cpp
class Solution {
public:
    int rec(int i, vector<int>& nums, vector<int>&memo, int &target) {
        int n = nums.size();
        if(i == n - 1) return 0;
        if(i == n) return INT_MIN;
        if(memo[i] != -2) return memo[i];
        
        int maxi = INT_MIN;

        for(int j = i + 1; j < n; j++) {
            int val = nums[j] - nums[i];
            if(-target <= val && val <= target) {
                maxi = max(maxi, 1 + rec(j, nums, memo, target));
            }
        }

        return memo[i] = maxi;
    } 
    int maximumJumps(vector<int>& nums, int target) {
        vector<int> memo(nums.size(), -2);
        int ans = rec(0, nums, memo, target);

        return ans < 0 ? -1 : ans;
    }
};